Proposal for multiple load of native programs. (#3566)
microsoft/ebpf-for-windows: docs/MultipleLoadNativeModule.md

1. ebpfcore will now maintain 2 hash tables to look up the native modules. One hash table has service name as key, and
another has module ID as key.
2. Currently native module is deleted from hash table when ref count becomes 0, which is before the driver is unloaded.
In this new design the native module should be deleted from the hash table only when the driver unload call has
completed. Also, we need to ensure that driver unload happens inline.

### Option 2
Another approach is to completely skip all the above synchronization logic in ebpfcore and ebpfapi.dll, and separate
out service install / uninstall completely from the scope of the application. Native modules will be installed by some
installer that installs the native module on the system. Once that is done, programs can be loaded / unloaded multiple
times, without the need of install / uninstall of service entries.

Changes needed to implement this:
1. Installer installs the service entry. We need to ensure the service name is unique. Two native modules with same
name but in different locations should be installed as 2 different services. Installer also populates the client module
ID registry key with a random GUID.

> 2. The app that loads the program will pass the service name to ebpfapi (TBD: How does app know the service name for its

Reviewer: This TBD could be addressed the same way as in lines 32-40, based on passing the file name in option 2.

Author: Removed this TBD

service). We can still use the existing `bpf_object__open` API, and the API can take the service name instead of file
name as input.

**Note**:
1. One limitation with option 2 is that all existing applications that are using native eBPF programs need to now

> also have an explicit installer / un-installer for the native programs that are they are using. This installer can

Reviewer: I disagree with this sentence.  My feedback was that the applications using the native program do NOT install the native program.  They only load it after it was installed by the same mechanism that got the native program on the machine in the first place.  E.g. that could be L3AFD or a script or whatever else.  It's also possible that the installer of the app and the installer of the native program are the same program, but they're logically two separate things which is why this sentence makes no sense to me.  An app installer can have a dependency on another components installer, just like multiple programs depend on say the .NET Framework being installed first, but that doesn't mean that multiple programs' installers all know the details of how to install the .NET Framework.  That's not how it works.

Author: @Alan-Jowett as per today's discussion in the triage meeting, for option 2, if the proposal is that ebpfapi exposes APIs for installers that helps install / uninstall the drivers, does it become similar to the approach in the below draft PR that adds 2 new APIs `ebpf_initialize_native_program_state()` and `ebpf_uninitialize_native_program_state()` for creating the service?

[#3556](https://github.com/microsoft/ebpf-for-windows/pull/3556/files#diff-8b744ef020b2dae7afefac77b8d1d9664fa3d249c683b307cd393362da9f5e8b)

Author: Updated doc with what was discussed in the triage meeting.

either be a separate entity (in the cases where the native eBPF program is a common program shared between multiple
applications / products), or it can be integrated with the application that loads the program (in the cases where the
native eBPF program is exclusively owned by that application / product).

> 2. Since this approach requires multiple components to generate the service name for the sys file, one option can be to

Reviewer: I disagree with this sentence too as noted in my comment on line 99.  How the service name can be generated based on the file path is completely orthogonal to whether to use option 1 or option 2.

# Load Multiple Instances of Native Module

> When using native mode on Windows, the programs are loaded from a native driver instead of an ELF file. As part of the load operation, the native driver is loaded in the kernel and the eBPF execution context orchestrates the creation of maps and loading the programs from the native driver. The execution context fetches the map properties from the native module (using functions exported by the native driver via NMR) and creates the required maps.

Reviewer: https://github.com/microsoft/ebpf-for-windows/blob/main/docs/DevelopmentGuide.md#formatting-notes says:
> Source lines **MUST NOT** exceed 120 columns.

Author: Fixed throughout the doc.

> But when trying to load the same native driver again, we hit a restriction that a driver cannot be loaded multiples times. This restriction creates a limitation where the same program(s) cannot be loaded multiple times when using native mode.

Reviewer: ```suggestion
But when trying to load the same native driver again, we hit a restriction that a driver cannot be loaded multiple times. This restriction creates a limitation where the same program(s) cannot be loaded multiple times when using native mode.
```

This document discusses proposal(s) for loading eBPF programs multiple times from the same native module.

At a high level, adding this feature support requires 2 changes:
1. Changes in ebpfapi and ebpfcore to synchronize service creation and deletion, along with driver load and unload.
1. Changes in BPF2C / native module to use "per-instance" map and helper contexts, instead of global contexts. Similarly, changes in ebpfcore to create the "per-instance" contexts when programs are loaded from native module, and pass the context to the program whenever the program is invoked.

This document mainly discusses the first change mentioned above, and possible options to implement it.

## Synchronize service creation
To ensure service creation / deletion, and driver load / unload are synchronized between multiple processes trying to load the same native module, following are the proposed options.

### Option 1

> 1. ebpfcore exposes a new IOCTL `GET_NATIVE_MODULE_HANDLE` which takes service name as input.

Reviewer: I agree in this option it should be the sys file name (without the .sys extension)

> 1. ebpfcore exposes a new IOCTL `GET_NATIVE_MODULE_HANDLE` which takes service name as input.

Author: So there are a couple of reasons:

1. ZwLoadDriver takes complete service path (format: `\\Registry\\Machine\\System\\CurrentControlSet\\Services\\<service_name>`) as parameter, so either ebpfcore or ebpfapi has to construct the complete service path. In the current implementation, ebpfapi creates the service (with name as a random GUID) and passes the complete path to ebpfcore.
2. Service name cannot be exactly same as the .sys file name. Consider a scenario where there are 2 different sys files (from 2 different products), located in different paths, but have same name. These 2 sys files should be allowed to be loaded at the same time. If the service name is same as sys file name, that will not work. To handle such a case, one option is to construct service path by taking file_name and appending it with 32-bit hash of the full path. That should make the service name unique.

> 1. ebpfcore exposes a new IOCTL `GET_NATIVE_MODULE_HANDLE` which takes service name as input.

Reviewer: > 2. Service name cannot be exactly same as the .sys file name. 

This conclusion doesn't follow from the example below.

> Consider a scenario where there are 2 different sys files (from 2 different products), located in different paths, but have same name. These 2 sys files should be allowed to be loaded at the same time. If the service name is same as sys file name, that will not work. To handle such a case, one option is to construct service path by taking file_name (without the extension) and appending it with 32-bit hash of the full path. That should make the service name unique.

I agree you need a unique service name.  But the developer of the native program, and the developer of any apps that want to load it, all need to know the unique identifier.  Methods to construct a unique identifier (of any sort) are:
a) Have a way to request/register a unique value in some central repository
b) Use a probabilistically random value as all or part of the identifier (the hash you mentioned is in this category)
c) Embed another existing identifier that has already been delegated to the developer, such as a domain name or Private Enterprise Number, where the owner of that identifier can ensure uniqueness of identifiers derived from it.

Once you have a way that developers use to construct a unique service name, they could name their .sys file based on that service name.  Hence the service name can use the .sys file name (plus the path prefix as you mention).

> 1. ebpfcore exposes a new IOCTL `GET_NATIVE_MODULE_HANDLE` which takes service name as input.

Author: Added more details in the doc about unique service name.

#### `GET_NATIVE_MODULE_HANDLE` Design:
1. UM will pass the service name. ebpfcore looks for the module based on the service name.
2. If ebpfcore does not find the module, it does the following:
    1. Generate a GUID for the client.
    2. Create a placeholder native module and add it to hash tables.
    3. Populate the service path (created by appending the service name to path prefix), and the module ID in the native module.
    4. Return the module HANDLE, module ID and return code `EBPF_SERVICE_NOT_PRESENT`.
3. If runtime finds the module, and it not in "unloading" state, it will return a new handle to module.
4. If runtime finds the module, and it is in "placeholder" state, it will block on an event, which will be set when the module is loaded.
5. If runtime finds the module, and it is in "unloading" state, it will block on another event, which will be set when call to driver unload returns. Once this event is set, this thread will be unblocked and move to step 2.

#### Misc. Changes

> 1. ebpfcore will now maintain 2 hash tables to lookup the native modules. One hash table has service name as key, and another has module ID as key.

Reviewer: ```suggestion
1. ebpfcore will now maintain 2 hash tables to look up the native modules. One hash table has service name as key, and another has module ID as key.
```

### Option 2
Another approach is to completely skip all the above synchronization logic in ebpfcore and ebpfapi.dll, and separate out service install / uninstall completely from the scope of the application. Native modules will be installed by some installer that installs the native module on the system. Once that is done, programs can be loaded / unloaded multiple times, without the need of install / uninstall of service entries.

Changes needed to implement this:
1. Installer installs the service entry. We need to ensure the service name is unique. Two native modules with same name but in different locations should be installed as 2 different services. Installer also populates the client module ID registry key with a random GUID.
2. The app that loads the program will pass the service name to ebpfapi (TBD: How does app know the service name for its service). We can still use the existing `bpf_object__open` API, and the API can take the service name instead of file name as input.

> **Note**: If we want to continue using existing APIs without any change, it causes a limitation that with this approach, **every user** of native module will need to implement the installer logic.

Reviewer: I didn't follow this sentence, can you elaborate?  What is a "user" here?
The native module installer (setup.exe or whatever) needs to implement it.  Calling apps don't, nor do their installers per se.

Author: I have updated (re-written) that note. If it still looks incorrect, I can remove that point altogether.

1. ebpfapi.dll provides two new APIs to install and uninstall service for a provided native module. These APIs take the
file name with optional path as input. Internally, ebpfapi generates a unique service name for provided file with path.
2. An installer installs the eBPF program on the machine by using the new eBPF service install APIs.
3. Once the service has been installed, an app can load the native module by passing the file name (with absolute or
relative path) to `bpf_object__open()`. ebpfapi.dll will use the logic as in step 1 above to internally regenerate the
same service name.
4. In the load step, ebpfapi.dll checks if the corresponding service is present. If it is present, it will proceed with
the load of program, else it will fail the load call.
5. In the case a program is shared between multiple apps / products, common installation will need to be managed by the
solutions that are using the shared eBPF program and is outside the scope of eBPF framework.
6. eBPF programs can be loaded multiple times from the same native module.
7. The native module will remain loaded until the service is uninstalled.

**Note**:

> One limitation with this options (option 2) is that all existing applications that are using native eBPF programs need

Reviewer: ```suggestion
One limitation with this option (option 2) is that all existing applications that are using native eBPF programs need
```
typo

to now also have an explicit installer / un-installer for the native programs that are they are using. This installer
can either be a separate entity (for example, in the cases where the native eBPF program is a common program shared

> between multiple applications / products), or it can be integrated with the application that loads the program (in the
> cases where the native eBPF program is exclusively owned by that application / product).

Reviewer: ```suggestion
between multiple applications / products), or it can be integrated with the application that loads the program (in the
cases where the native eBPF program is exclusively owned by that application / product, and the application / product runs as admin).
```
I don't think we should encourage having an app/product running as admin, so I would prefer to delete the whole ", or it can be ..." to the end of the sentence.
